design-proposals/bound-token-for-vnc.md

```diff
@@ -0,0 +1,77 @@
+# Overview
+
+Kubernetes API server can create 
+[tokens](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens)
+bound to a service account. They are mainly used in services running inside the cluster to authenticate
+between each other. 
+
+This document describes how to create a token that can be used to authenticate with the API server
+to only access VNC endpoint of a singe VM.
+
+## Motivation
+
+An administrator may want to give VNC access to a client application without giving it permissions to access
+anything else in the cluster. The VNC client could be pointed to VMI `/vnc` subresource and use
+the token to authenticate. 
+
+## Goals
+
+- A temporary token can be created to access VNC.
+- It can only be used with one specific VM.
+- It is valid only for a specific duration.
+
+## User Stories
+
+- A VM owner wants to give access to VNC to a third party application without giving it permissions to modify the VM.
+
+## Repos
+
+- https://github.com/kubevirt/kubevirt
+
+# Design
+
+An API subresource under `virtualmachineinstance` , `/vnc-token` is used to get a token for VNC.
+
+The token will refer to a `ServiceAccount` that only has permissions to access the `/vnc` subresource on
+one specific VM.
+
+
+## API Examples
+
+Getting the VNC token:
+```
+GET /apis/subresources.kubevirt.io/v1/namespaces/${VMI_NAMESPACE}/virtualmachineinstances/${VMI_NAME}/vnc-token
+```
+
+Parameters:
+- `duration` - Time duration while the token is valid.
+
+
+## Scalability
+
+Creating a token requires an API call to the kubernetes API server.
+
+## Update/Rollback Compatibility
+
+### Update
+A new version of `virt-controller` will create and watch `ServiceAccounts`, `Roles` and `RoleBindings` for each VMI
```

Comment:
Why pre-create these resources, couldn't we do it per request?

Reply:
Thanks but that's still a pre-created `ServiceAccount` per` VirtualMachineInstance` right? Why not only create the `ServiceAccount` with the first token request?

Reply:
Apologies as I'm likely missing something here but couldn't the reconcile logic be applied by the controller once the resource had been created by the api? Why does this only work if created by the controller? 

I don't want to bike shed on this point it just seems odd to pre-create `ServiceAccounts` that given the use case aren't going to be used by the majority of `VirtualMachineInstances`.

Reply:
There is one big concern which I have here.

Just grabbing our `virt-controller` cluster-role:

```yaml
rules:
- apiGroups:
  - policy
  resources:
  - poddisruptionbudgets
  verbs:
  - get
  - list
  - watch
  - delete
  - create
  - patch
- apiGroups:
  - ""
  resources:
  - pods
  - configmaps
  - endpoints
  verbs:
  - get
  - list
  - watch
  - delete
  - update
  - create
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - update
  - create
  - patch
- apiGroups:
  - ""
  resources:
  - pods/finalizers
  verbs:
  - update
- apiGroups:
  - ""
  resources:
  - pods/eviction
  verbs:
  - create
- apiGroups:
  - ""
  resources:
  - pods/status
  verbs:
  - patch
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
  - list
  - watch
  - update
  - patch
- apiGroups:
  - apps
  resources:
  - daemonsets
  verbs:
  - list
- apiGroups:
  - apps
  resources:
  - controllerrevisions
  verbs:
  - watch
  - list
  - create
  - delete
- apiGroups:
  - ""
  resources:
  - persistentvolumeclaims
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - delete
  - patch
- apiGroups:
  - snapshot.kubevirt.io
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - pool.kubevirt.io
  resources:
  - virtualmachinepools
  - virtualmachinepools/finalizers
  verbs:
  - watch
  - list
  - create
  - delete
  - update
  - patch
  - get
- apiGroups:
  - kubevirt.io
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - subresources.kubevirt.io
  resources:
  - virtualmachineinstances/addvolume
  - virtualmachineinstances/removevolume
  - virtualmachineinstances/freeze
  - virtualmachineinstances/unfreeze
  - virtualmachineinstances/softreboot
  verbs:
  - update
- apiGroups:
  - cdi.kubevirt.io
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - k8s.cni.cncf.io
  resources:
  - network-attachment-definitions
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - authorization.k8s.io
  resources:
  - subjectaccessreviews
  verbs:
  - create
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshotclasses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshots
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - delete
- apiGroups:
  - storage.k8s.io
  resources:
  - storageclasses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - flavor.kubevirt.io
  resources:
  - virtualmachineflavors
  - virtualmachineclusterflavors
  - virtualmachinepreferences
  - virtualmachineclusterpreferences
  verbs:
  - list
  - watch
- apiGroups:
  - migrations.kubevirt.io
  resources:
  - migrationpolicies
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - namespaces
  verbs:
  - get
```

you can see that we don't need access in any form to ServiceAccounts`, `Role` or `RoleBinding`. What we basically allow the virt-controller SA now, is giving any service account in any namespace the same amount of permissions which it has itself.

I would as an alternative consider the following design:

 * e.g. maintainer temporary in-memory tokens in virt-handler, which can be created and retrieved over `vnc-token`
 * add a vnc endpoint which can be accesses by a SA, hand out these sa credentials  to any user
 * in the vnc endpoint, when reaching virt-handler, validate that the token is valid

This would not require more privileges. Maybe other options which don't require privilege increase, exist too.

Reply:
What do you think about this idea:
- VMI will have a special annotation (for example `vnc-token.kubevirt.io: true`) that will signal to `virt-controller` that it should create and watch the `ServiceAccount` and other resources for this VMI. This annotation will not be there by default.
- On request to `/vnc-token` endpoint, `virt-api` will add this annotation, then wait until the `ServiceAccount` is created and create a token.

Reply:
Hm, not sure that helps, since virt-controller would still need permissions to service accounts and so forth. With the possible alternaitve, e.g. virt-operator can create this permanent service account with basically no permissions. Just to avoid DOS on a possible token-based endpoint. If DOS is not a concern (like apiserver would only be by accident public accessible, and there still the token would prevent active misuse), we may not even need that.

Reply:
What I try to see, is if there is a way to provide this feature without increasing virt-controller privileges.

Reply:
I have removed my previous comment, because I forgot to refresh this page before I posted it and it referred to Lee's comment.

@rmohr, how can we pass the token used for authentication with k8s API into `virt-handler` to check if it is correct?
Can you maybe point me to some place in the code, or explain how could I implement it?

Reply:
Or how about this idea. The token has `audiences` field which could contain the UID of the VMI that it can access.

https://github.com/kubevirt/kubevirt/blob/692d3801a89413a417e1d7553e469f4010b6779a/vendor/k8s.io/api/authentication/v1/types.go#L159

Then `virt-api` on receiving `/vnc` request, could check if the UID from the token is equal to the UID of the VMI. This means that we don't need to create and watch `ServiceAccounts` for every VMI. There would be a single `ServiceAccount` with only access to `/vnc` and from it all tokens would be created.

Now I need to figure out how to access the token audiences from `virt-api`.

Reply:
@akrejcir just pushed an experiment I did a while ago to https://github.com/rmohr/kubevirt/pull/19. It roughly implements what I mean. Let me know what you think.

Reply:
I see, thanks.

I couldn't find a way to access the bound ServiceAccount token that is used for authentication from `virt-api`, so my idea in the previous comment is not possible.

So I think we can use your idea, to generate random tokens in `virt-handler`. I will change the proposal.

Reply:
So, you don't just have to take the proposal. There are some points to clarify:

 * for convenience it would be best, if this extra REST endpoint is reachable without auth just with the token, is that ok from a security perspective?
 * if the VM migrates, a new token would have to be generated. So sometimes you may have the issue that you ask for a token which is valid for 30 seconds, but it is already invalid after 1 second.
 
 I would therefore list it as an alternative implementation and we can discuss both, instead of just replacing your proposal.
The limitations above may not be much of an issue but they really depend on the actual usage-pattern which we want to provide.

Like, you may want to hand out tokens which are valid for one week. That would not be possible. If you want to just hand out via a user-portal temporary access tokens which are valid for 1 minute, so that retrieval and usage can function properly (e.g. what oVirt does with the qemu tokens), then we are very likely good.

Reply:
I have pushed a new change with three design alternatives.